File: IOS/Packages/LiquorEngine/Sources/LiquorEngine/Passport.swift
import Foundation

/// The passport: the distilleries you have stood in.
///
/// A visit is a date and a name. Read back, the visits become stamps --
/// one per distillery, with how many times and when -- and two lists the
/// shelf can answer that no passport can: which of your bottles came from
/// a place you have been, and which distilleries on your shelf you have
/// never visited. Nothing here is a checklist of the catalogue; the
/// distilleries you have not been to are the ones whose bottles you own,
/// which is a list you might act on, not a list you cannot finish.
public enum Passport: Sendable {

    public struct Visit: Hashable, Sendable {
        public let distillery: String
        public let at: Date
        public let note: String?

        public init(distillery: String, at: Date, note: String? = nil) {
            self.distillery = distillery; self.at = at; self.note = note
        }
    }

    /// A bottle, as the passport needs it: where it was made and where it
    /// was bought.
    public struct Bottle: Hashable, Sendable {
        public let name: String
        public let distillery: String?
        public let boughtAt: String?

        public init(name: String, distillery: String?, boughtAt: String? = nil) {
            self.name = name; self.distillery = distillery; self.boughtAt = boughtAt
        }
    }

    public struct Stamp: Hashable, Sendable, Identifiable {
        public var id: String { key }
        public let key: String
        /// The spelling you used most recently.
        public let name: String
        public let visits: Int
        public let firstAt: Date
        public let lastAt: Date
        /// Bottles on the shelf made there.
        public let bottlesFromThere: [String]
        /// Bottles whose purchase store is the distillery itself.
        public let boughtThere: [String]
    }

    public struct Summary: Hashable, Sendable {
        /// Most recently visited first.
        public let stamps: [Stamp]
        public let visits: Int
        /// Distilleries with bottles on the shelf and no visit, most
        /// bottles first.
        public let notYetVisited: [(name: String, bottles: Int)]
        /// "6 distilleries, 9 visits." Nil with nothing recorded.
        public let headline: String?

        public static func == (a: Summary, b: Summary) -> Bool {
            a.stamps == b.stamps && a.visits == b.visits && a.headline == b.headline
                && a.notYetVisited.map(\.name) == b.notYetVisited.map(\.name)
        }
        public func hash(into hasher: inout Hasher) { hasher.combine(stamps); hasher.combine(visits) }
    }

    public static func summarise(visits: [Visit], shelf: [Bottle]) -> Summary {
        var byKey: [String: [Visit]] = [:]
        for v in visits {
            let key = normalise(v.distillery)
            guard !key.isEmpty else { continue }
            byKey[key, default: []].append(v)
        }
        let stamps = byKey.map { key, rows -> Stamp in
            let newest = rows.max { $0.at < $1.at }!
            let made = shelf.filter { $0.distillery.map(normalise) == key }.map(\.name).sorted()
            let bought = shelf.filter { $0.boughtAt.map(normalise) == key }.map(\.name).sorted()
            return Stamp(
                key: key, name: newest.distillery.trimmingCharacters(in: .whitespaces),
                visits: rows.count,
                firstAt: rows.map(\.at).min()!, lastAt: newest.at,
                bottlesFromThere: made, boughtThere: bought)
        }
        .sorted { $0.lastAt != $1.lastAt ? $0.lastAt > $1.lastAt : $0.name < $1.name }

        var unvisited: [String: (name: String, bottles: Int)] = [:]
        for bottle in shelf {
            guard let distillery = bottle.distillery else { continue }
            let key = normalise(distillery)
            guard !key.isEmpty, byKey[key] == nil else { continue }
            unvisited[key] = (name: distillery, bottles: (unvisited[key]?.bottles ?? 0) + 1)
        }
        let notYet = unvisited.values.sorted { $0.bottles != $1.bottles ? $0.bottles > $1.bottles : $0.name < $1.name }

        let headline: String? = stamps.isEmpty ? nil
            : "\(stamps.count) \(stamps.count == 1 ? "distillery" : "distilleries"), \(visits.count) \(visits.count == 1 ? "visit" : "visits")."
        return Summary(stamps: stamps, visits: visits.count, notYetVisited: notYet, headline: headline)
    }

    /// "3 visits · first March 2024 · last 3 weeks ago" or "Once, 3 weeks ago".
    public static func line(_ s: Stamp, now: Date = Date(), calendar: Calendar = .current) -> String {
        let last = Hunt.ago(AgeMath.days(from: s.lastAt, to: now, calendar: calendar))
        if s.visits == 1 { return "Once, \(last)" }
        return "\(s.visits) visits · first \(monthAndYear(s.firstAt, calendar: calendar)) · last \(last)"
    }

    /// "2 bottles from there on the shelf, 1 bought there." Nil with neither.
    public static func shelfLine(_ s: Stamp) -> String? {
        var parts: [String] = []
        if !s.bottlesFromThere.isEmpty {
            parts.append("\(s.bottlesFromThere.count) \(s.bottlesFromThere.count == 1 ? "bottle" : "bottles") from there on the shelf")
        }
        if !s.boughtThere.isEmpty {
            parts.append("\(s.boughtThere.count) bought there")
        }
        return parts.isEmpty ? nil : parts.joined(separator: ", ") + "."
    }

    public static func normalise(_ name: String) -> String {
        name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            .replacingOccurrences(of: " distillery", with: "")
            .replacingOccurrences(of: " distilling co.", with: "")
            .replacingOccurrences(of: " distilling company", with: "")
            .replacingOccurrences(of: " distilling", with: "")
    }

    static func monthAndYear(_ date: Date, calendar: Calendar) -> String {
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.timeZone = calendar.timeZone
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "MMMM yyyy"
        return formatter.string(from: date)
    }
}

File: Android/engine/src/main/kotlin/com/talastack/liquorlog/engine/Passport.kt
package com.talastack.liquorlog.engine

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * The passport: the distilleries you have stood in.
 *
 * A visit is a date and a name. Read back, the visits become stamps -- one
 * per distillery, with how many times and when -- and two lists the shelf
 * can answer that no passport can: which of your bottles came from a place
 * you have been, and which distilleries on your shelf you have never
 * visited. Nothing here is a checklist of the catalogue; the distilleries
 * you have not been to are the ones whose bottles you own, which is a list
 * you might act on, not a list you cannot finish.
 */
object Passport {

    data class Visit(
        val distillery: String,
        val at: Instant,
        val note: String? = null,
    )

    /** A bottle, as the passport needs it: where it was made and where it was bought. */
    data class Bottle(
        val name: String,
        val distillery: String?,
        val boughtAt: String? = null,
    )

    data class Stamp(
        val key: String,
        /** The spelling you used most recently. */
        val name: String,
        val visits: Int,
        val firstAt: Instant,
        val lastAt: Instant,
        /** Bottles on the shelf made there. */
        val bottlesFromThere: List<String>,
        /** Bottles whose purchase store is the distillery itself. */
        val boughtThere: List<String>,
    )

    data class Unvisited(val name: String, val bottles: Int)

    data class Summary(
        /** Most recently visited first. */
        val stamps: List<Stamp>,
        val visits: Int,
        /**
         * Distilleries with bottles on the shelf and no visit, most bottles
         * first.
         */
        val notYetVisited: List<Unvisited>,
        /** "6 distilleries, 9 visits." Null with nothing recorded. */
        val headline: String?,
    )

    fun summarise(visits: List<Visit>, shelf: List<Bottle>): Summary {
        val byKey = visits
            .filter { normalise(it.distillery).isNotEmpty() }
            .groupBy { normalise(it.distillery) }

        val stamps = byKey.map { (key, rows) ->
            val newest = rows.maxByOrNull { it.at }!!
            Stamp(
                key = key,
                name = newest.distillery.trim(),
                visits = rows.size,
                firstAt = rows.minOf { it.at },
                lastAt = newest.at,
                bottlesFromThere = shelf
                    .filter { it.distillery?.let(::normalise) == key }
                    .map { it.name }.sorted(),
                boughtThere = shelf
                    .filter { it.boughtAt?.let(::normalise) == key }
                    .map { it.name }.sorted(),
            )
        }.sortedWith(compareByDescending<Stamp> { it.lastAt }.thenBy { it.name })

        // LinkedHashMap so a tie in bottle count resolves by name, not by
        // whatever order the shelf happened to arrive in.
        val unvisited = LinkedHashMap<String, Unvisited>()
        for (bottle in shelf) {
            val distillery = bottle.distillery ?: continue
            val key = normalise(distillery)
            if (key.isEmpty() || byKey.containsKey(key)) continue
            // The latest spelling wins, as it does for a stamp: the name
            // shown is the one most recently typed.
            unvisited[key] = Unvisited(
                name = distillery,
                bottles = (unvisited[key]?.bottles ?: 0) + 1,
            )
        }
        val notYet = unvisited.values
            .sortedWith(compareByDescending<Unvisited> { it.bottles }.thenBy { it.name })

        val headline = if (stamps.isEmpty()) {
            null
        } else {
            "${stamps.size} ${if (stamps.size == 1) "distillery" else "distilleries"}, " +
                "${visits.size} ${if (visits.size == 1) "visit" else "visits"}."
        }
        return Summary(stamps = stamps, visits = visits.size, notYetVisited = notYet, headline = headline)
    }

    /** "3 visits · first March 2024 · last 3 weeks ago" or "Once, 3 weeks ago". */
    fun line(s: Stamp, now: Instant = Instant.now(), zone: ZoneId = ZoneId.systemDefault()): String {
        val last = Hunt.ago(AgeMath.days(s.lastAt, now))
        if (s.visits == 1) return "Once, $last"
        return "${s.visits} visits · first ${monthAndYear(s.firstAt, zone)} · last $last"
    }

    /** "2 bottles from there on the shelf, 1 bought there." Null with neither. */
    fun shelfLine(s: Stamp): String? {
        val parts = mutableListOf<String>()
        if (s.bottlesFromThere.isNotEmpty()) {
            parts += "${s.bottlesFromThere.size} " +
                "${if (s.bottlesFromThere.size == 1) "bottle" else "bottles"} from there on the shelf"
        }
        if (s.boughtThere.isNotEmpty()) {
            parts += "${s.boughtThere.size} bought there"
        }
        return if (parts.isEmpty()) null else parts.joinToString(", ") + "."
    }

    /**
     * "Buffalo Trace Distillery" and "buffalo trace" are one place. The
     * suffixes are stripped after lowercasing, so the order of these
     * replacements is the order they appear on a label.
     */
    fun normalise(name: String): String =
        name.trim().lowercase()
            .replace(" distillery", "")
            .replace(" distilling co.", "")
            .replace(" distilling company", "")
            .replace(" distilling", "")

    private fun monthAndYear(date: Instant, zone: ZoneId): String =
        DateTimeFormatter.ofPattern("MMMM yyyy", Locale.US).withZone(zone).format(date)
}
